lib/rules/jest/no-if.js

```diff
@@ -17,69 +17,54 @@ module.exports = {
         'or not testing what you intend to.',
         'Consider breaking the if statement out',
         'into a separate test to resolve this error.',
+        'If your conditionals are required to',
+        'satisfy the typescript type checker, consider',
+        'using a non-null assertion operator (!) instead.',
       ].join(' '),
       noConditional: [
         'Tests should not contain conditional statements.',
         'This is usually an indication that you',
         'are attempting to test too much at once',
         'or not testing what you intend to.',
-        'Consider breaking the conditional statement out',
-        'into a separate test to resolve this error.',
+        'Consider writing a separate test for',
+        'each fork in the conditional statement.',
+        'If your conditionals are required to',
+        'satisfy the typescript type checker, consider',
+        'using a non-null assertion operator (!) instead.',
       ].join(' '),
     },
   },
 
   create(context) {
-    let callExpressionDepth = 0;
+    const stack = [];
 
-    return {
-      CallExpression(node) {
-        if (ignore(node)) {
-          return;
-        }
-        callExpressionDepth++;
-      },
-      IfStatement(node) {
-        if (callExpressionDepth === 0) {
-          return;
-        }
+    function validate(node) {
+      const lastCallExpression = stack[stack.length - 1];
 
-        context.report({
-          messageId: 'noIf',
-          node,
-        });
-      },
-      ConditionalExpression(node) {
-        if (callExpressionDepth === 0) {
-          return;
-        }
+      if (!lastCallExpression) {
+        return;
+      }
+
+      if (isTestDefinition(lastCallExpression)) {
```

Comment:
Is there a reason to compute this manually every time? Can the stack just have the results of calling this?